sunpy/net/scraper_utils.py

```diff
@@ -0,0 +1,139 @@
+                   'millisecond': timedelta(milliseconds=1)}
+
+def check_timerange(pattern, url, timerange):
+    """
+    Checks whether the time range represented in *url* intersects
+    with the given time range.
+
+    Parameters
+    ----------
+    url : `str`
+        URL of the file.
+    timerange : `~sunpy.time.TimeRange`
+        Time interval for which files were searched.
+
+    Returns
+    -------
+    `bool`
+        `True` if URL's valid time range overlaps the given timerange, else `False`.
+    """
+    exdict = parse(pattern, url).named
+    if exdict['year'] < 100:
+        exdict['year'] = 2000 + exdict['year']
+    if 'month' not in exdict:
+                if 'month_name' in exdict:
+                    exdict['month'] = datetime.strptime(exdict['month_name'], '%B').month
+                elif 'month_name_abbr' in exdict:
+                    exdict['month'] = datetime.strptime(exdict['month_name_abbr'], '%b').month
+    tr = get_timerange_from_exdict(exdict)
+    return tr.intersects(timerange)
+
+def smaller_pattern(directoryPattern):
+    """
+    Obtain the smaller time step for the given pattern.
+    """
+    if "%S" in directoryPattern:
+        return relativedelta(seconds=1)
+    elif "%M" in directoryPattern:
+        return relativedelta(minutes=1)
+    elif any(hour in directoryPattern for hour in ["%H"]):
+        return relativedelta(hours=1)
+    elif any(day in directoryPattern for day in ["%d", "%j"]):
+        return relativedelta(days=1)
+    elif any(month in directoryPattern for month in ["%b", "%B", "%m"]):
+        return relativedelta(months=1)
+    elif any(year in directoryPattern for year in ["%Y", "%y"]):
+        return relativedelta(years=1)
+    else:
+        return None
+
+def date_floor(date, timestep):
+    """
+    Return the "floor" of the given date and time step.
+
+    Parameters
+    ----------
+    datetime : `datetime.datetime` or `astropy.time.Time`
+        The date to floor
+    timestep : `dateutil.relativedelta.relativedelta`
+        The smallest time step to floor
+    Returns
+    -------
+    `datetime.datetime`
+        The time floored at the given time step
+    """
+    date_parts = [int(p) for p in date.strftime('%Y,%m,%d,%H,%M,%S').split(',')]
+    date_parts[-1] = date_parts[-1] % 60
+    date = datetime(*date_parts)
+    orig_time_tup = date.timetuple()
+    time_tup = [orig_time_tup.tm_year, orig_time_tup.tm_mon, orig_time_tup.tm_mday,
+                orig_time_tup.tm_hour, orig_time_tup.tm_min, orig_time_tup.tm_sec]
+    if timestep == relativedelta(minutes=1):
+        time_tup[-1] = 0
+    elif timestep == relativedelta(hours=1):
+        time_tup[-2:] = [0, 0]
+    elif timestep == relativedelta(days=1):
+        time_tup[-3:] = [0, 0, 0]
+    elif timestep == relativedelta(months=1):
+        time_tup[-4:] = [1, 0, 0, 0]
+    elif timestep == relativedelta(years=1):
+        time_tup[-5:] = [1, 1, 0, 0, 0]
```

Comment:
There seems to be a unit test coverage problem here, can you confirm and if so, add one please?